### PR TITLE
Use `schema_name` to create the `physical_name`

### PR DESCRIPTION
**Pull request by joroKr21**
More consistency and less opportunity for column name mismatch.

## Which issue does this PR close?

Part of #11782.

## Rationale for this change

It's not manageable to have these two slightly different ways of obtaining a name from an expression.
Also, minor differences in the implementation can cause schema mismatch errors.

## What changes are included in this PR?

`physical_name` delegates to `schema_name` for all cases except top-level columns which are unqualified.
This was previously the purpose of the `is_first_expr` flag in `create_physical_name`.

## Are these changes tested?

Relying on existing tests.

## Are there any user-facing changes?

`Expr::canonical_name` is deprecated because it's just an alias for `format!`

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -1104,6 +1103,7 @@ impl Expr {
     }
 
     /// Returns a full and complete string representation of this expression.
+    #[deprecated(note = "use format! instead")]
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:1106]: 👍

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -2386,263 +2386,13 @@ fn fmt_function(
...
-        Expr::Placeholder(_) => {
-            internal_err!("Create physical name does not support placeholder")
-        }
-        Expr::OuterReferenceColumn(_, _) => {
-            internal_err!("Create physical name does not support OuterReferenceColumn")
-        }
+/// The name of the column (field) that this `Expr` will produce in the physical plan.
+/// The difference from [Expr::schema_name] is that top-level columns are unqualified.
+pub fn physical_name(expr: &Expr) -> Result<String> {
+    if let Expr::Column(col) = expr {
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:2392]: 😍

File: datafusion/physical-plan/src/aggregates/mod.rs [hunk 3]
```diff
@@ -2179,6 +2179,7 @@ mod tests {
             .map(|order_by_expr| {
                 let ordering_req = order_by_expr.unwrap_or_default();
                 AggregateExprBuilder::new(array_agg_udaf(), vec![Arc::clone(col_a)])
+                    .alias("a")
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/mod.rs:2182]: why is this change needed?

**Review by alamb — APPROVED** [review]
Thank you @joroKr21  -- this is an amazing PR (amazing to delete so much code but not change any tests -- that really shows a deep understanding of the code)

cc @mustafasrepo and @jayzhan211  as I vaguely remember we discussed the need for this alternate path for display expressions. Clearly as all the tests pass the second copy isn't needed it seems

**Comment by jayzhan211** [comment]
We should get physical name from Physical expression not from `Expr`. If we are building the expr from builder, there is no `Expr` that you could get the name from.
Btw, if you are interesting to solve this issue, let me know, so I can work on other task

File: datafusion/physical-plan/src/aggregates/mod.rs (hunk 3 above)
Comment (joroKr21) [reply datafusion/physical-plan/src/aggregates/mod.rs:2182]: Do we really need it? Is there a situation where we would not provide the alias to the physical expression?
I made this change because the alias generated previously was incorrect (it didn't use the arguments).

**Comment by joroKr21** [comment]
> If we are building the expr from builder, there is no Expr that you could get the name from.

Is there a use case for this? The way I see it, physical expressions are low-level enough that you would have to explicitly provide an alias. But I'm curious to know.

**Comment by jayzhan211** [comment]
> > If we are building the expr from builder, there is no Expr that you could get the name from.
> 
> Is there a use case for this? The way I see it, physical expressions are low-level enough that you would have to explicitly provide an alias. But I'm curious to know.

In datafusion-comet, they heavily rely on physical layer stuffs. They directly build the physical expression with out logical expression.

https://github.com/apache/datafusion/issues/11359
This is also the reason why logical expression confused them while it is unnecessary for them to build physical expression.

**Comment by alamb** [comment]
Maybe @andygrove / @viirya  could also provide some feedback -- I don't understand the complexities of the naming requirements in comet

If there are some other non obvious requirements for comet, it would be great to get some additional tests in DataFusion that demonstrate the usecase so that we don't break something for comet accidentally.

File: datafusion/expr/src/expr.rs [hunk 4]
```diff
@@ -2386,263 +2386,13 @@ fn fmt_function(
...
-            None => Ok(format!("*{}", options)),
-        },
-        Expr::Placeholder(_) => {
-            internal_err!("Create physical name does not support placeholder")
-        }
-        Expr::OuterReferenceColumn(_, _) => {
-            internal_err!("Create physical name does not support OuterReferenceColumn")
-        }
+/// The name of the column (field) that this `Expr` will produce in the physical plan.
+/// The difference from [Expr::schema_name] is that top-level columns are unqualified.
```
Comment (viirya) [inline datafusion/expr/src/expr.rs:2390]: It would be better if we can provide example like `schema_name` to show difference.

**Comment by viirya** [comment]
> This was previously the purpose of the is_first_expression flag in create_physical_name.

I don't see the appearance of `is_first_expression` flag in the diff. So the description looks a bit confused.

**Comment by viirya** [comment]
> Maybe @andygrove / @viirya could also provide some feedback -- I don't understand the complexities of the naming requirements in comet
> 
> If there are some other non obvious requirements for comet, it would be great to get some additional tests in DataFusion that demonstrate the usecase so that we don't break something for comet accidentally.

As Comet doesn't go through the query analysis of DataFusion, I think it should be fine from name resolution change.

File: datafusion/expr/src/expr.rs [hunk 5]
```diff
@@ -2386,263 +2386,13 @@ fn fmt_function(
...
-    Ok(order_by
-        .map(|order_by| format!("{} ORDER BY [{}]", phys_name, expr_vec_fmt!(order_by)))
-        .unwrap_or(phys_name))
-}
-
-pub fn physical_name(e: &Expr) -> Result<String> {
-    create_physical_name(e, true)
-}
-
-fn create_physical_name(e: &Expr, is_first_expr: bool) -> Result<String> {
```
Comment (joroKr21) [inline datafusion/expr/src/expr.rs:2416]: @viirya this was the flag, I remembered it incorrectly: `is_first_expr`

**Review by viirya — COMMENTED** [review]
It looks good to have a consistent approach for physical name for expression, instead of both `physical_name` and `schema_name`.

**Comment by alamb** [comment]
What is the current status of this PR?

I am a little confused

@viirya  seems to suggest this would be a good change, but @jayzhan211 points out this change would mean anyone who built `PhysicalExpr`s without a corresponding `Expr` would have to create their own schema names

> We should get physical name from Physical expression not from Expr. If we are building the expr from builder, there is no Expr that you could get the name from.

So I guess my question is "if we merge this PR would it mess up comet (or other users of only physical exprs)?"

**Comment by viirya** [comment]
> @viirya seems to suggest this would be a good change, but @jayzhan211 points out this change would mean anyone who built `PhysicalExpr`s without a corresponding `Expr` would have to create their own schema names

Hmm, as we built physical operators directly (without going through analysis stage in DataFusion), we manually bind all columns. Do we still have additional name resolution in physical stage in DataFusion?

**Comment by jayzhan211** [comment]
> Do we still have additional name resolution in physical stage in DataFusion?

The name is used for projection mapping and also the display in explain statement
https://github.com/apache/datafusion/blob/67cf1d606b901eaf591f44389702e285557d39e1/datafusion/physical-expr/src/equivalence/projection.rs#L66-L74

**Comment by viirya** [comment]
It looks like a simple check that if a column's name matches corresponding input field name in input schema. Isn't? As we provide matched input schema and column names to projection, I think it should be all good. I'm not sure how does this PR impact us.

**Comment by jayzhan211** [comment]
>  I'm not sure how does this PR impact us.

It will not. Merge this PR does not impact Comet.

> Do we still have additional name resolution in physical stage in DataFusion?

The response for this is just to inform we still have name resolution in physical stage in Datafusion so the name of physical expressions matters and might not be a good idea to get it from the logical expression directly.

**Comment by alamb** [comment]
Ok, I'll plan to merge this PR tomorrow unless anyone else has a different opinion

Thank you all for the discussion. I am glad we got to a good place
